### ISSUE TITLE
Covariance matrix assembly in CGAL is very unefficient

### ISSUE THREAD

**Issue by sgiraudot** [CGAL/cgal#2069]
## Issue Details

I just noticed that applying `CGAL::linear_least_squares_fitting_3` to a big set of triangles was suspiciously long. Using _Valgrind_, I saw that most of the time was spent on `new` and `delete` operators. It seems that the internal code to assemble the covariance matrix uses `CGAL::Linear_algebraCd` were all matrices are dynamically allocated (and as one matrix is computed per triangle and multiplied, the penalty is high), see in `Principal_component_analysis/include/CGAL/PCA_util.h`.

I tried to replace the few lines of linear algebra using the `Eigen` library (matrices are static) and, well, it's about __200 times faster__ (yes, 200 times, from 2 minutes to 0.6 seconds on my test). I'm obviously going to make a proper branch and submit a PR, but I'm not sure how we want to do this:

- PCA is an old package that can use Eigen (`Eigen_diagonalize_traits`) but also has a fallback mode using internal diagonalization code (`Diagonalize_traits`)
- For backward compatibility, we should keep this fallback mode
- My first idea was to specialize the functions (when `DiagonalizeTraits` is `Eigen_diagonalize_traits`) but there are already many functions and code duplicate (for handling points, triangles, cuboids, etc.). Sadly, `CGAL::Linear_algebraCd` and `Eigen` do not have the exact same API (matrix access is `m[][]` in CGAL and `m(,)` in Eigen) so we can't do a minimal `#ifdef/#else` somewhere
- Maybe add a layer of abstraction just to bind the few linear algebra operations either to `CGAL::Linear_algebraCd` and `Eigen`?

Also, I'm not sure where `CGAL::Linear_algebraCd` is used in CGAL, but it seems to me that in many cases, matrices have a known size at compilation time, so it could also be a good idea to provide matrices with template size parameters to avoid the numerous allocation/disallocation. But that's probably more complicated than that, so maybe not a good idea to change this code.

__Note:__ just for the record, this problem is not present when using PCA on a set of points (there is no linear algebra involved here), which is the most widely used case.

**Comment by lrineau** [comment]
`Linear_algebraCd` is used mainly in the old kernel `Carterian_d` (and relatives), in Jet_fitting, and PCA:
```shell
renoir ~/Git/cgal $ ack Linear_algebraCd ^build*/(include|examples|demo) -l
Jet_fitting_3/include/CGAL/Monge_via_jet_fitting.h
Kernel_d/include/CGAL/Cartesian_d.h
Kernel_d/include/CGAL/Kernel_d/Interval_linear_algebra.h
Kernel_d/include/CGAL/Kernel_d/Linear_algebraCd_impl.h
Kernel_d/include/CGAL/Linear_algebraCd.h
Kernel_d/include/CGAL/Filtered_kernel_d.h
Principal_component_analysis/include/CGAL/linear_least_squares_fitting_triangles_2.h
Principal_component_analysis/include/CGAL/PCA_util.h
Principal_component_analysis/include/CGAL/linear_least_squares_fitting_rectangles_2.h
Principal_component_analysis/include/CGAL/linear_least_squares_fitting_segments_2.h
Principal_component_analysis/include/CGAL/linear_least_squares_fitting_circles_2.h
```

**Comment by afabri** [comment]
I would switch to Eigen even if it creates a dependency, but @palliez should make the final decision

**Comment by lrineau** [comment]
Fixed by #2074.
